[editing] Add offset range check in CompositeEditCommand::replaceSelectedTextInNode().

ziransun opened the pull request:
#### 4672c550c929439b3b291808899763bfc439e4ec
<pre>
[editing] Add offset range check in CompositeEditCommand::replaceSelectedTextInNode().

<a href="https://bugs.webkit.org/show_bug.cgi?id=150969">https://bugs.webkit.org/show_bug.cgi?id=150969</a>

Reviewed by Ryosuke Niwa and Darin Adler.

When replacing text in a text node, we need to valid the offset
range against the length of the text node. For example, in the
fixed test, the length of text node is 0, in which case, there is
no need to call replaceTextInNode() function.

* LayoutTests/TestExpectations:
* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::doApply):

Canonical link: <a href="https://commits.webkit.org/284343@main">https://commits.webkit.org/284343@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cfcce6a23a1b9659404973b9aea44ecf719014d3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69020 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48420 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21692 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73100 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20178 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71137 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20026 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54959 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13404 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72086 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44202 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59586 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35433 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40867 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17015 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18552 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62816 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74809 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13002 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16607 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13040 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62508 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15336 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10486 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4093 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44224 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45297 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46493 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45039 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -135,6 +135,21 @@ void InsertTextCommand::doApply()
+    if (endingSelection().start().anchorType() == Position::PositionIsOffsetInAnchor) {
+        const unsigned startOffset = endingSelection().start().offsetInContainerNode();
+        auto startNode = dynamicDowncast<CharacterData>(endingSelection().start().anchorNode());
+        if (startNode && startOffset > startNode->length())
```
Why don't we clamp the value to length instead of bailing out completely.

rniwa commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -135,6 +135,21 @@ void InsertTextCommand::doApply()
+    if (endingSelection().end().anchorType() == Position::PositionIsOffsetInAnchor) {
+        const unsigned endOffset = endingSelection().end().offsetInContainerNode();
+        auto endNode = dynamicDowncast<CharacterData>(endingSelection().end().anchorNode());
+        if (endNode && endOffset > endNode->length())
```
Ditto.

darinadler reviewed (commented):
Clamping is a good idea, and probably better than dumping the insert operation. But I’m not sure it solves the whole problem. Why do we end up with out-of-range offsets? We’d want to fix the cause of that, too, not just clamp.

darinadler commented on Source/WebCore/editing/CompositeEditCommand.cpp:
```diff
@@ -802,6 +802,8 @@ Position CompositeEditCommand::replaceSelectedTextInNode(const String& text)
         return Position();
 
     RefPtr<Text> textNode = start.containerText();
+    if (start.offsetInContainerNode() >= static_cast<int>(textNode->length()) || end.offsetInContainerNode() > static_cast<int>(textNode->length()))
```
This looks roughly correct. Not sure we have enough test coverage, though.

Does this function ever get used to replace a zero-length selection, like putting text at an insertion point? If so, then we should allow the start offset to be equal to the length, and it should be > rather than >=.

There seem to be some missing checks here. For example, no check that `end.offsetInContainerNode() >= start.offsetInContainerNode()`. But maybe I’m misunderstanding what the foundation is for needing each check.

darinadler reviewed (approved):
Bug needs a new title since it specifically says something we are not doing.

ziransun committed:
[editing] Add offset range check in CompositeEditCommand::replaceSelectedTextInNode().

https://bugs.webkit.org/show_bug.cgi?id=150969

Reviewed by Ryosuke Niwa and Darin Adler.

When replacing text in a text node, we need to valid the offset
range against the length of the text node. For example, in the
fixed test, the length of text node is 0, in which case, there is
no need to call replaceTextInNode() function.

* LayoutTests/TestExpectations:
* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::doApply):

Canonical link: https://commits.webkit.org/284343@main
